### PR TITLE
Implemented Pillar Pixel Map Construction Method

### PR DESCRIPTION
**Pull request by adisingh50**
Implemented method: `construct_pillar_pixel_map()` in `forward_projector.py`. 
Also created a unit test for it called `test_construct_pillar_pixel_map()`.
- Also, I moved the "Approach 1" method, `sample_ground_points_argoverse()` out of `model.py` and into its own utility file, since this approach 1 is not really specific to the `bev_mapping` module

Once this is finished, then I'll have all the functionality in place to create a visual unit test for map_constructor.

### CODE REVIEW

File: rasternet/utils/ground_point_utils.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+import pdb
+
+import numpy as np
+import torch
+import torch.nn as nn
+from argoverse.map_representation.map_api import ArgoverseMap
+from argoverse.utils.se3 import SE3
+
+
+def sample_ground_points_argoverse(
```
Comment (adisingh50) [inline rasternet/utils/ground_point_utils.py:14]: this "approach 1" method already existed from before. I just moved it into its own file

File: tests/bev_mapping/test_forward_projector.py [hunk 2]
```diff
@@ -82,5 +76,60 @@ def test_forward_projection_shape():
...
+
+    dataset_dir = CALIB_FILE = "/srv/share2/aahluwalia30/argoverse-tracking/all_logs"
+    log_id = "6f153f9c-edc5-389f-ac6f-40705c30d97e"
+    CALIB_FILE_PATH = f"{dataset_dir}/{log_id}/vehicle_calibration_info.json"
+    calib = load_calib(CALIB_FILE_PATH)
+    NUM_SAMPLES_PER_PILLAR = 2
+    forwardProjector = ForwardProjector(calib["ring_front_center"], NUM_SAMPLES_PER_PILLAR)
+
+    img_height = forwardProjector.img_height  # 1200
+    img_width = forwardProjector.img_width  # 1900
```
Comment (anshulcgm) [inline tests/bev_mapping/test_forward_projector.py:102]: img_width is 1920

File: rasternet/bev_mapping/forward_projector.py [hunk 3]
```diff
@@ -87,26 +91,49 @@ def project_egovehicle_coords_to_img(self, sampled_pillar_points: np.ndarray) ->
...
+        # Iteratively populate pillar_pixel_map with 1s for any valid correspondences
+        # between pixels and pillars.
+        pillar_pixel_map = np.zeros((num_pillars, num_pts, self.img_height * self.img_width))
+        for pillar_index in range(num_pillars):
+            for point_index in range(num_pts):
+                pixel_coord = real_img_coords[pillar_index, point_index, :]
+
+                u_coord = int(pixel_coord[0])
+                v_coord = int(pixel_coord[1])
+
```
Comment (anshulcgm) [inline rasternet/bev_mapping/forward_projector.py:129]: Here you also need to make sure the z-coordinate is not behind the camera, reference the determine_valid_cam_coords method to figure out how to do that. You need to take the points in the camera coordinate frame and the points in pixel coordinates and then make sure the z-values of the camera coordinate frame are not negative and the points in the pixel coordinates fit within the image resolution.

File: tests/bev_mapping/test_forward_projector.py [hunk 4]
```diff
@@ -82,5 +76,60 @@ def test_forward_projection_shape():
...
+
+    # fmt: off
+    # testing 4 pillars, each containing 2 points
```
Comment (anshulcgm) [inline tests/bev_mapping/test_forward_projector.py:105]: Add tests to check for negative z-coordinates

File: rasternet/bev_mapping/forward_projector.py (hunk 3 above)
Comment (adisingh50) [reply rasternet/bev_mapping/forward_projector.py:129]: got it.

File: rasternet/bev_mapping/forward_projector.py [hunk 5]
```diff
@@ -87,26 +91,49 @@ def project_egovehicle_coords_to_img(self, sampled_pillar_points: np.ndarray) ->
...
+        # Obtain true pixel coordinates by dividing by homogeneous coordinates.
+        hom_coords = image_plane_pts[:, :, 2].reshape(num_pillars, num_pts, 1)
+        real_img_coords = (image_plane_pts / hom_coords)[:, :, :2]
+
+        # Iteratively populate pillar_pixel_map with 1s for any valid correspondences
+        # between pixels and pillars.
+        pillar_pixel_map = np.zeros((num_pillars, num_pts, self.img_height * self.img_width))
+        for pillar_index in range(num_pillars):
+            for point_index in range(num_pts):
```
Comment (johnwlambert) [inline rasternet/bev_mapping/forward_projector.py:124]: for millions of points, this for-loop could be quite slow. I believe it can be totally vectorized.

Comment (adisingh50) [reply rasternet/bev_mapping/forward_projector.py:132]: Good point. I'm a bit stuck on efficiently vectorizing this process. But @anshulcgm says he has an idea of how to do it. So we are going to merge this branch in (with the inefficienty for-loop implementation), and then he will vectorize it in the next PR.

File: rasternet/bev_mapping/forward_projector.py [hunk 6]
```diff
@@ -58,27 +59,27 @@ def sample_pillar_points(self, ground_pillar_pts: np.ndarray) -> np.ndarray:
 
         return sampled_pillar_points
 
-    def determine_valid_cam_coords(uv: np.ndarray, uv_cam: np.ndarray, camera_config: CameraConfig) -> np.ndarray:
+    def determine_valid_point_projections(self, image_plane_pts: np.ndarray, cam_coordinates: np.ndarray) -> np.ndarray:
```
Comment (adisingh50) [inline rasternet/bev_mapping/forward_projector.py:62]: modified this function so that @anshulcgm can call it from `construct_pillar_pixel_map()` when he goes for the vectorized implementation

File: rasternet/bev_mapping/forward_projector.py (hunk 5 above)
Comment (johnwlambert) [reply rasternet/bev_mapping/forward_projector.py:132]: @adisingh50 gotcha, ok.

@adisingh50 @anshulcgm my idea was that you could project all points in the beginning, and then identify the corresponding pillar indices after the fact by just dividing the original 3d coordinates by the grid size, and rounding down.

File: rasternet/bev_mapping/forward_projector.py [hunk 7]
```diff
@@ -87,26 +92,58 @@ def project_egovehicle_coords_to_img(self, sampled_pillar_points: np.ndarray) ->
...
+                v_coord = int(pixel_coord[1])
+                cam_z_coord = cam_coord[2]
+
+                # Don't populate with 1 if the (u,v) coordinate is invalid or camera's z coordinate is negative.
+                u_valid = (0 <= u_coord) and (u_coord < self.img_width)
+                v_valid = (0 <= v_coord) and (v_coord < self.img_height)
+                if (not u_valid) or (not v_valid) or (cam_z_coord < 0):
+                    continue
+
+                flattened_img_coord = (v_coord * self.img_width) + u_coord
```
Comment (johnwlambert) [inline rasternet/bev_mapping/forward_projector.py:146]: didn't quite follow the intuition behind the math for this line

Comment (adisingh50) [reply rasternet/bev_mapping/forward_projector.py:146]: Basically, I'm given a (u,v) coordinate of the 1920x1200 image and then i'm calculating what the 1d coordinate is if that image were completely flattened into a 1d array (so of length 1920*1200 = 2,304,000). I need this 1d coordinate in order to mark a correspondence between a specific pillar and a specific image pixel index. This article kind of reiterates what I'm trying to do: https://softwareengineering.stackexchange.com/questions/212808/treating-a-1d-data-structure-as-2d-grid

**Comment by adisingh50** [comment]
Merging this in now, so that Anshul can vectorize the `construct_pillar_pixel_map()` on his branch. I will now start working on the visual unit test for map_constructor (now that pillar_pixel_map is completed, although inefficient).
